File: javascript/packages/config/src/framework-detection.ts
import path from "path"

import { promises as fs } from "fs"

import type { Framework, PrismNode, PrismParseResult } from "@herb-tools/core"

export const GEMFILE_NAMES = ["Gemfile", "gems.rb"]

export const FRAMEWORK_GEMS: Record<string, Framework> = {
  rails: "actionview",
  actionview: "actionview",
  hanami: "hanami",
  "hanami-view": "hanami",
  sinatra: "sinatra",
}

const FRAMEWORK_PRECEDENCE: Framework[] = ["actionview", "hanami", "sinatra"]

export const FRAMEWORK_LABELS: Record<Framework, string> = {
  ruby: "Ruby",
  actionview: "Action View",
  hanami: "Hanami",
  sinatra: "Sinatra",
}

export const FRAMEWORK_DESCRIPTIONS: Record<Framework, string> = {
  ruby: "plain ERB templates, with no framework helpers in scope",
  actionview: "Action View templates, with their helpers, partials, and strict locals",
  hanami: "Hanami views, with their parts and helpers",
  sinatra: "Sinatra templates, with their helpers",
}

export interface RubySourceParser {
  parseRuby(source: string): PrismParseResult
}

export interface FrameworkDetection {
  framework: Framework
  gem: string
  gemfilePath: string
}

function collectGems(node: PrismNode, gems: string[]): void {
  if (!node || typeof node.childNodes !== "function") return

  if (node.constructor?.name === "CallNode" && node.name === "gem" && !node.receiver) {
    const [firstArgument] = node.arguments_?.arguments_ ?? []
    const name = gemName(firstArgument)

    if (name) {
      gems.push(name)
    }
  }

  for (const child of node.childNodes()) {
    collectGems(child, gems)
  }
}

function gemName(node: PrismNode | undefined): string | undefined {
  const type = node?.constructor?.name

  if (type !== "StringNode" && type !== "SymbolNode") return undefined

  return node.unescaped?.value
}

export function gemsFromGemfile(source: string, parser: RubySourceParser): string[] {
  let result: PrismParseResult

  try {
    result = parser.parseRuby(source)
  } catch {
    return []
  }

  const program = result.value

  if (!program) return []

  const gems: string[] = []

  collectGems(program, gems)

  return gems
}

export async function detectFrameworkFromGemfile(
  projectPath: string,
  parser: RubySourceParser
): Promise<FrameworkDetection | undefined> {
  for (const gemfileName of GEMFILE_NAMES) {
    const gemfilePath = path.join(projectPath, gemfileName)

    let source: string

    try {
      source = await fs.readFile(gemfilePath, "utf8")
    } catch {
      continue
    }

    const gems = gemsFromGemfile(source, parser)
    const frameworks = new Map<Framework, string>()

    for (const gem of gems) {
      const framework = FRAMEWORK_GEMS[gem]

      if (framework && !frameworks.has(framework)) {
        frameworks.set(framework, gem)
      }
    }

    for (const framework of FRAMEWORK_PRECEDENCE) {
      const gem = frameworks.get(framework)

      if (gem) {
        return { framework, gem, gemfilePath }
      }
    }

    return undefined
  }

  return undefined
}

File: rust/herb-config/src/framework_detection.rs
use std::path::{Path, PathBuf};

use crate::config_schema::Framework;

pub const GEMFILE_NAMES: [&str; 2] = ["Gemfile", "gems.rb"];

pub const FRAMEWORK_GEMS: [(&str, Framework); 5] = [
  ("rails", Framework::ActionView),
  ("actionview", Framework::ActionView),
  ("hanami", Framework::Hanami),
  ("hanami-view", Framework::Hanami),
  ("sinatra", Framework::Sinatra),
];

const FRAMEWORK_PRECEDENCE: [Framework; 3] = [Framework::ActionView, Framework::Hanami, Framework::Sinatra];

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FrameworkDetection {
  pub framework: Framework,
  pub gem: String,
  pub gemfile_path: PathBuf,
}

pub fn framework_for_gem(gem: &str) -> Option<Framework> {
  FRAMEWORK_GEMS.iter().find(|(name, _)| *name == gem).map(|(_, framework)| *framework)
}

pub fn framework_from_gems(gems: &[String]) -> Option<(Framework, String)> {
  FRAMEWORK_PRECEDENCE.iter().find_map(|framework| {
    gems
      .iter()
      .find(|gem| framework_for_gem(gem) == Some(*framework))
      .map(|gem| (*framework, gem.clone()))
  })
}

#[cfg(feature = "prism")]
pub fn gems_from_gemfile(source: &str) -> Vec<String> {
  let mut gems = Vec::new();

  let Ok(result) = herb::parse_ruby(source) else {
    return gems;
  };

  let Some(program) = result.program() else {
    return gems;
  };

  collect_gems(&program, &mut gems);

  gems
}

#[cfg(not(feature = "prism"))]
pub fn gems_from_gemfile(_source: &str) -> Vec<String> {
  Vec::new()
}

#[cfg(feature = "prism")]
fn collect_gems(node: &herb::prism::PrismNode, gems: &mut Vec<String>) {
  if node.is("CallNode") && node.name.as_deref() == Some("gem") && node.receiver().is_none() {
    if let Some(gem) = gem_name(node) {
      gems.push(gem);
    }
  }

  for child in &node.children {
    collect_gems(child, gems);
  }
}

#[cfg(feature = "prism")]
fn gem_name(call: &herb::prism::PrismNode) -> Option<String> {
  let arguments = call.children.iter().find(|child| child.is("ArgumentsNode"))?;
  let first_argument = arguments.children.first()?;

  if first_argument.is("StringNode") || first_argument.is("SymbolNode") {
    first_argument.unescaped.clone()
  } else {
    None
  }
}

pub fn detect_framework_from_gemfile(project_path: &Path) -> Option<FrameworkDetection> {
  for gemfile_name in GEMFILE_NAMES {
    let gemfile_path = project_path.join(gemfile_name);

    let Ok(source) = std::fs::read_to_string(&gemfile_path) else {
      continue;
    };

    let (framework, gem) = framework_from_gems(&gems_from_gemfile(&source))?;

    return Some(FrameworkDetection { framework, gem, gemfile_path });
  }

  None
}
